### Data Merge

##### Declarations

In [1]:
#Importing utilities file with library imports and helper functions
%run "utils.ipynb"

In [2]:
#Loading input pickle files
df_income_1901_edited=pd.read_pickle(input_path+"\\df_income_1901_edited.pkl")
df_health_outcomes_edited=pd.read_pickle(input_path+"\\df_health_outcomes_edited.pkl")
df_sdoh_edited=pd.read_pickle(input_path+"\\df_sdoh_edited.pkl")
df_state_politics_edited=pd.read_pickle(input_path+"\\df_state_politics_edited.pkl")
df_state_public_spend_edited=pd.read_pickle(input_path+"\\df_state_public_spend_edited.pkl")
df_acs_edited=pd.read_pickle(input_path+"\\df_acs_edited.pkl")
df_zip2st_edited=pd.read_pickle(input_path+"\\df_zip2st_edited.pkl")
df_urban_rural_edited=pd.read_pickle(input_path+"\\df_urban_rural_edited.pkl")

In [3]:
#Printing shapes to check which file needs most attention for data compression
for df in [df_income_1901_edited, df_health_outcomes_edited, df_sdoh_edited, df_state_politics_edited, 
           df_state_public_spend_edited, df_acs_edited, df_urban_rural_edited]:
    print(df.shape)

(33774, 15)
(1169992, 10)
(291024, 4)
(57, 3)
(255, 3)
(33120, 14)
(33178, 2)


##### ZIP level data

In [4]:
[df_income_1901_edited.ZIP.nunique(), df_health_outcomes_edited.ZIP.nunique(), df_sdoh_edited.ZIP.nunique(), 
 df_acs_edited.ZIP.nunique(), df_urban_rural_edited.ZIP.nunique()]

[33774, 32409, 32336, 33120, 33178]

In [5]:
df_income_1901_essential=df_income_1901_edited[['ZIP', 'Estimate Households Total', 'Percent Households lt 10k', 
                                                'Estimate Households Median income (dollars)', 'Estimate Families Total',
                                                'Estimate Families Median income (dollars)', 'Estimate Nonfamily households Total',
                                                'Estimate Nonfamily households Median income (dollars)', 'Percent Households that are Families']]
df_income_1901_essential.shape

(33774, 9)

In [6]:
df_health_outcomes_essential=df_health_outcomes_edited[['ZIP', 'TotalPopulation', 'Short_Question_Text', 'Data_Value']]
df_health_outcomes_essential_wide=df_health_outcomes_essential.pivot(index=['ZIP', 'TotalPopulation'], columns='Short_Question_Text', 
                                                                     values='Data_Value').reset_index().rename_axis(None, axis=1)
#display(df_health_outcomes_essential_wide.head())
df_health_outcomes_essential_wide.shape

(32409, 39)

In [7]:
df_sdoh_essential_wide=df_sdoh_edited.pivot(index=['ZIP', 'TotalPopulation'], columns='Measure', 
                                                                     values='Data_Value').reset_index().rename_axis(None, axis=1)
#display(df_sdoh_essential_wide.head())
df_sdoh_essential_wide.shape

(32336, 11)

##### State level data

In [8]:
#Checking if any of the state level data needs to be pivoted
print("df_state_politics_edited", df_state_politics_edited.shape, df_state_politics_edited.State.nunique())
print("df_zip2st_edited", df_zip2st_edited.shape, df_zip2st_edited.ZIP.nunique())
print("df_state_public_spend_edited", df_state_public_spend_edited.shape, df_state_public_spend_edited['State Name'].nunique())

df_state_politics_edited (57, 3) 57
df_zip2st_edited (39368, 2) 39368
df_state_public_spend_edited (255, 3) 51


In [9]:
#Turns out state_public_spend_edited will need to be pivoted as it has multiple values of year for each state
df_state_public_spend_essential_wide=df_state_public_spend_edited.pivot(
    index=['State Name'], columns='TimeFrame', values='Data').reset_index().rename_axis(
        None, axis=1).add_prefix('spend_').rename(columns={'spend_State Name':'State Name'})
df_state_public_spend_essential_wide.head()

State Name  spend_2017  spend_2018  spend_2019  spend_2020  spend_2021
0     Alabama    56.43214       57.23       54.25       46.81       52.35
1      Alaska   114.43666       96.52       98.25      215.15       90.17
2     Arizona     8.83289        9.66       15.25       24.34       15.32
3    Arkansas    52.29329       50.57       48.23       45.90       42.94
4  California    61.69589       65.71       71.89       69.87       78.47

In [10]:
#Geographic information (Lat-Long)
geocoords=df_health_outcomes_edited[['ZIP', 'longitude', 'latitude']].drop_duplicates()
geocoords.shape, geocoords.ZIP.nunique()

((32409, 3), 32409)

##### Creating combined dataset

In [54]:
# Combined_dataset_zip contains ZIP level features.
# Combined_dataset_state contains State level features.
# Combined_dataset contains all features

# Merging dataframes for ZIP level features
combined_dataset_zip = pd.merge(
    pd.merge(
        pd.merge(
            pd.merge(
                df_income_1901_essential,
                df_health_outcomes_essential_wide,
                on="ZIP",
                how="outer",
            ),
            df_sdoh_essential_wide,
            on="ZIP",
            how="outer",
        ),
        df_acs_edited,
        on="ZIP",
        how="outer",
    ),
    df_urban_rural_edited,
    on="ZIP",
    how="left",
)

# Merging dataframes for State level features
combined_dataset_state = pd.merge(
    pd.merge(
        df_state_politics_edited,
        df_state_public_spend_essential_wide,
        on="State Name",
        how="outer",
    ),
    df_zip2st_edited,
    on="State",
    how="outer",
)

# Merging ZIP and State level features
combined_dataset = pd.merge(
    pd.merge(combined_dataset_zip, combined_dataset_state, on="ZIP", how="outer"),
    geocoords,
    on="ZIP",
    how="left",
)

# Print the shape of the combined dataset
combined_dataset.shape

(39503, 81)

In [55]:
combined_dataset.sample(5)

ZIP  Estimate Households Total  Percent Households lt 10k  \
18146  53190                     6962.0                       12.7   
7809   24894                      125.0                       15.2   
36344  35270                        NaN                        NaN   
28170  79372                      546.0                        3.7   
8297   26287                     1148.0                        5.9   

       Estimate Households Median income (dollars)  Estimate Families Total  \
18146                                      56923.0                   3342.0   
7809                                           NaN                     74.0   
36344                                          NaN                      NaN   
28170                                      61563.0                    430.0   
8297                                       53021.0                    748.0   

       Estimate Families Median income (dollars)  \
18146                                    83022.0   
7809                                     21563.0   
36344                                        NaN   
28170                                    74688.0   
8297                                     68667.0   

       Estimate Nonfamily households Total  \
18146                               3620.0   
7809                                  51.0   
36344                                  NaN   
28170                                116.0   
8297                                 400.0   

       Estimate Nonfamily households Median income (dollars)  \
18146                                            32059.0       
7809                                                 NaN       
36344                                                NaN       
28170                                            23182.0       
8297                                             28676.0       

       Percent Households that are Families  TotalPopulation_x  \
18146                              0.480034            19200.0   
7809                               0.592000              388.0   
36344                                   NaN                NaN   
28170                              0.787546             1434.0   
8297                               0.651568             3131.0   

       All Teeth Lost  Annual Checkup  Any Disability  Arthritis  \
18146            13.0            70.4            26.1       18.2   
7809             37.3            78.1            52.5       40.5   
36344             NaN             NaN             NaN        NaN   
28170            16.8            69.1            32.8       23.9   
8297             20.8            80.0            38.5       35.9   

       Binge Drinking  COPD  Cancer (except skin)  Cervical Cancer Screening  \
18146            25.0   4.8                   4.6                       74.3   
7809             11.1  18.8                   7.1                       76.1   
36344             NaN   NaN                   NaN                        NaN   
28170            17.9   7.6                   6.0                       78.3   
8297             12.4  11.6                   8.2                       82.9   

       Cholesterol Screening  Chronic Kidney Disease  Cognitive Disability  \
18146                   65.6                     2.1                  16.7   
7809                    81.0                     4.2                  28.2   
36344                    NaN                     NaN                   NaN   
28170                   82.2                     3.1                  15.7   
8297                    87.8                     3.6                  17.1   

       Colorectal Cancer Screening  Core preventive services for older men  \
18146                         59.2                                    38.7   
7809                          59.1                                    36.7   
36344                          NaN                                     NaN   
28170                         61.0                                    

In [56]:
combined_dataset['longitude'].isnull().sum()

7094

In [57]:
#Saving combined dataset
combined_dataset.to_pickle(output_path+"\\combined_dataset_anuvrat.pkl")
combined_dataset.to_csv(output_path+"\\combined_dataset_anuvrat.csv")

##### Checking if combined_dataset has values for all ZIPs

In [58]:
# % Missing values for each column. Function defined in helper functions notebook
perc_missing_num(combined_dataset)

Number of records in dataset: 39503
Missing records summary:


column  number_missing  \
0                                                 ZIP               3   
71                                         State Name             132   
72                                              State             132   
73  Political Affiliation (2008-2020 presidential ...             319   
74                                         spend_2017             319   
76                                         spend_2019             319   
75                                         spend_2018             319   
77                                         spend_2020             319   
78                                         spend_2021            2399   
4                             Estimate Families Total            5729   
1                           Estimate Households Total            5729   
6                 Estimate Nonfamily households Total            5729   
58  Estimate!!SEX AND AGE!!Total population!!Sex r...            6383   
62  Percent Margin of Error!!RACE!!Total populatio...            6383   
57            Estimate!!SEX AND AGE!!Total population            6383   
59                    Percent!!RACE!!Total population            6383   
60          Percent!!RACE!!Total population!!One race            6383   
61  Percent!!RACE!!Total population!!One race!!Bla...            6383   
63  Percent!!RACE!!Total population!!One race!!Ame...            6383   
69                      Estimate!!Total housing units            6383   
65  Percent!!RACE!!Total population!!One race!!Nat...            6383   
66  Percent Margin of Error!!RACE!!Total populatio...            6383   
67  Percent!!HISPANIC OR LATINO AND RACE!!Total po...            6383   
68  Percent Margin of Error!!HISPANIC OR LATINO AN...            6383   
64  Percent Margin of Error!!RACE!!Total populatio...            6383   
8                Percent Households that are Families            6685   
2                           Percent Households lt 10k            6685   
43                                     Sleep <7 hours            7094   
79                                          longitude            7094   
80                                           latitude            7094   
9                                   TotalPopulation_x            7094   
27                                       Dental Visit            7094   
17                          Cervical Cancer Screening            7103   
52                     Persons aged 65 years or older            7167   
54        Persons of racial or ethnic minority status            7167   
47                                  TotalPopulation_y            7167   
21                        Colorectal Cancer Screening            7194   
51  No high school diploma among adults aged 25 ye...            7208   
36                                        Mammography            7253   
10                                     All Teeth Lost            7273   
22             Core preventive services for older men            7315   
56  Unemployment among people 16 years and older i...            7319   
23           Core preventive services for older women            7322   
53     Persons living below 150% of the poverty level            7460   
49               Housing cost burden among households            7504   
55                           Single-parent households            7504   
50  No broadband internet subscription among house...            7504   
48                       Crowding among housing units            7504   
70                                      Percent_Urban            7971   
19                             Chronic Kidney Disease            8068   
18                              Cholesterol Screening            8068   
16                               Cancer (except skin)            8068   
15                                               COPD            8068   
14                                     Binge Drinking            8068   
13                              

##### Dataset with strict removal of any missing value (except % Urban)

In [59]:
# Giving -99 values to missing urban_percentage as we don't want to drop ZIPs just because of this column
combined_dataset_per_urban_99=combined_dataset.copy()
combined_dataset_per_urban_99['Percent_Urban'].where(combined_dataset_per_urban_99['Percent_Urban'].notnull(), -99, inplace=True)

In [60]:
# Creating the dataset
combined_dataset_nona_strict=combined_dataset_per_urban_99.dropna()

In [61]:
# % Missing values for each column. Function defined in helper functions notebook
#perc_missing_num(combined_dataset_nona_strict)

In [62]:
combined_dataset_nona_strict.shape

(24089, 81)

#### Checking which states are completely absent in the new dataset

In [66]:
# List of states in original dataset and dataset without any missing information
states_in_original=combined_dataset['State Name'].unique()
states_in_nona_strict=combined_dataset_nona_strict['State Name'].unique()

In [67]:
print("States not present in nona but present in original dataset")
for state in states_in_original:
    if state not in states_in_nona_strict: print(state)

States not present in nona but present in original dataset
Puerto Rico
nan
Rhode Island
Delaware
West Virginia
Florida
Kansas
Utah
American Samoa
Northern Mariana Islands
Trust Territories
Guam
Virgin Islands


In [69]:
#combined_dataset[combined_dataset['State Name'].isin(['Kansas', 'Utah', 'Florida', 'Delaware', 'West Virginia'])].sample(20)
print("# unique values for spend 2021 for Utah, Kansas, Delaware, and West Virginia:", 
      combined_dataset[combined_dataset['State Name'].isin(['Utah', 'Kansas', 'Delaware', 'West Virginia'])]['spend_2021'].unique())
print("# unique values for Annual Checkup for Florida:", 
      combined_dataset[combined_dataset['State Name'].isin(['Florida'])]['Annual Checkup'].unique())

# Note: Kansas, Utah, Delaware, and West Virginia does not have Spend 2021 variable populated. 
## Note2: Florida doesn't have majority of health outcomes populated.


# unique values for spend 2021 for Utah, Kansas, Delaware, and West Virginia: [nan]
# unique values for Annual Checkup for Florida: [nan]


In [70]:
# It seems resonable to drop 'spend_2021' to get Kansas, Utah, Delaware, and West Virginia back in the data. 
# However, I can't see a path to include Florida. 
# We can probably take it as a separate case and verify our findings from rest of the data.
combined_dataset_nona=combined_dataset.drop(columns=['spend_2021']).dropna()

In [71]:
# % Missing values for each column. Function defined in helper functions notebook
#perc_missing_num(combined_dataset_nona)

In [72]:
states_in_nona=combined_dataset_nona['State Name'].unique()
print("States not present in nona but present in original dataset")
for state in states_in_original:
    if state not in states_in_nona: print(state)

States not present in nona but present in original dataset
Puerto Rico
nan
Florida
American Samoa
Northern Mariana Islands
Trust Territories
Guam
Virgin Islands


In [73]:
#Summary of #ZIPs by State to see national coverage. Unfortunately Florida is not included
combined_dataset_nona[['State Name', 'ZIP']].groupby(by='State Name').count()

ZIP
State Name                
Alabama                511
Alaska                 138
Arizona                289
Arkansas               408
California            1365
Colorado               368
Connecticut            244
Delaware                53
District of Columbia    21
Georgia                569
Hawaii                  72
Idaho                  195
Illinois              1164
Indiana                643
Iowa                   815
Kansas                 568
Kentucky               475
Louisiana              377
Maine                  343
Maryland               354
Massachusetts          469
Michigan               892
Minnesota              806
Mississippi            312
Missouri               807
Montana                235
Nebraska               476
Nevada                 108
New Hampshire          199
New Jersey             517
New Mexico             169
New York              1385
North Carolina         667
North Dakota           266
Ohio                   995
Oklahoma               552
Oregon                 314
Pennsylvania          1375
Rhode Island            61
South Carolina         334
South Dakota           255
Tennessee              526
Texas                 1471
Utah                   196
Vermont                215
Virginia               654
Washington             453
West Virginia          339
Wisconsin              724
Wyoming                 86

In [75]:
combined_dataset_nona_florida=combined_dataset[combined_dataset['State Name'].isin(['Florida'])].dropna(axis=1, how='all').dropna()
combined_dataset_nona_florida.shape

(839, 52)

Insights:
<ol>
<li>29 health outcomes features don't have data for 8,068 or 20% ZIPs. Most of these ZIPs should be 0 or low population as per Census procedure documented in the notes for this dataset.</li>
<li>Another 21 features don't have data for at least 7,000 or more ZIPs.</li>
<li>Another 18 features don't have data for at least 5,704 or more ZIPs.</li>
<li>Removing all ZIPs with missing data leaves us with 24,089 ZIPs (61%) out of 39,503 ZIPs in the original combined dataset. But it will exclude Kansas, Utah and Florida all together.</li>
<li>Kansas, Utah, Delaware and West Virginia do not have Spend 2021 variable populated. Florida doesn't have majority of health outcomes populated.</li>
<li>Recommend to  drop spend 2021 to get Kansas, Utah, Delaware and West Virginia back in the data. </li>
<li>However, I can't see a path to include Florida. We can probably take it as a separate case and verify our findings from rest of the data.</li>
<li>This leaves us with 25,321 ZIPs with 80 features in the primary dataset and 893 ZIPs with 51 features in Florida dataset. </li>
<li>This seems like a reasonable sample size with national coverage. Dropping ZIPs with missing data will remove the noise.</li>
<li>Recommend to move ahead with 25K ZIPs for further analysis.</li>
</ol>

##### Saving dataset without any missing information for any ZIP

In [76]:
#Saving dataset without any missing information
combined_dataset_nona.to_pickle(output_path+"\\combined_dataset_nona.pkl")
combined_dataset_nona.to_csv(output_path+"\\combined_dataset_nona.csv")

#Saving separate dataset for Florida
combined_dataset_nona_florida.to_pickle(output_path+"\\combined_dataset_nona_florida.pkl")
combined_dataset_nona_florida.to_csv(output_path+"\\combined_dataset_nona_florida.csv")